external-crates/move/solana/move-to-solana/src/stackless/translate.rs

```diff
@@ -215,6 +218,55 @@ impl<'mm, 'up> FunctionContext<'mm, 'up> {
         self.env.module_env.env
     }
 
+    // Note for two following routines.
+    // Stackless code may have basic block terminators being removed by Move compiler optimizer,
+    // LLVM requires all basic block terminators even for the fall-thru.
+
+    // This is a general approach to fix all basic blocks which do not have a terminator.
+    // The only requirement for this function is that 'next' and 'prev' iterators for the basic blocks
+    // should return the next and prev bb in the code layout (and not in the control flow).
+    // This is the right assumption, since we build basic blocks by sequencial processing of the labels
+    // and instructions.
+    pub fn fix_bb_terminators(&self) {
+        let function = self
+            .module_cx
+            .lookup_move_fn_decl(self.env.get_qualified_inst_id(self.type_params.to_vec()))
+            .0;
+
+        let mut current_block = unsafe { LLVMGetEntryBasicBlock(function) };
+        loop {
+            let next_block = unsafe { LLVMGetNextBasicBlock(current_block) };
+            if next_block.is_null() {
+                break;
+            }
+            let no_terminator = unsafe { LLVMGetBasicBlockTerminator(current_block).is_null() };
+            if no_terminator {
+                unsafe {
+                    LLVMPositionBuilderAtEnd(self.module_cx.llvm_builder.0, current_block);
+                    LLVMBuildBr(self.module_cx.llvm_builder.0, next_block);
+                }
+            }
+
+            current_block = next_block;
+        }
+    }
+
+    // Check and add terminator in the previous basic block.
+    // This is handy to use when basic blocks are already built and while translating the Label bytecode.
+    fn fix_terminator_in_prev_bb(&self, llbb: BasicBlock) {
+        let bb = *llbb.get_basic_block_ref();
+        unsafe {
+            let prev_bb = LLVMGetPreviousBasicBlock(bb);
+            if !prev_bb.is_null() {
+                let terminator = LLVMGetBasicBlockTerminator(prev_bb);
+                if terminator.is_null() {
+                    LLVMPositionBuilderAtEnd(self.module_cx.llvm_builder.0, prev_bb);
+                    LLVMBuildBr(self.module_cx.llvm_builder.0, bb);
```

Comment:
Yes, we know that LLVMGetPreviousBasicBlock gives the correct predecessor since it is how we build them in `translate_instruction`.
Providing an additional tracing will be just an unnecessary double booking.